mllib/src/main/scala/org/apache/spark/ml/classification/MultinomialLogisticRegression.scala

```diff
@@ -0,0 +1,622 @@
+                } else {
+                  0.0
+                }
+              }
+            }
+          }
+          new BreezeOWLQN[Int, BDV[Double]]($(maxIter), 10, regParamL1Fun, $(tol))
+        }
+
+        val initialCoefficientsWithIntercept = Vectors.zeros(numClasses * numFeaturesPlusIntercept)
+
+        if ($(fitIntercept)) {
+          /*
+             For multinomial logistic regression, when we initialize the coefficients as zeros,
+             it will converge faster if we initialize the intercepts such that
+             it follows the distribution of the labels.
+             {{{
+               P(0) = \exp(b_0) / (\sum_{k=1}^K \exp(b_k))
+               ...
+               P(K) = \exp(b_K) / (\sum_{k=1}^K \exp(b_k))
+             }}}
+             The solution to this is not identifiable, so choose the solution with minimum
+             L2 penalty (i.e. subtract the mean). Hence,
+             {{{
+               b_k = \log{count_k / count_0}
+               b_k' = b_k - \frac{1}{K} \sum b_k
+             }}}
+           */
+          val referenceCoef = histogram.indices.map { i =>
+            if (histogram(i) > 0) {
+              math.log(histogram(i) / (histogram(0) + 1)) // add 1 for smoothing
+            } else {
+              0.0
+            }
+          }
+          val referenceMean = referenceCoef.sum / referenceCoef.length
+          histogram.indices.foreach { i =>
+            initialCoefficientsWithIntercept.toArray(i * numFeaturesPlusIntercept + numFeatures) =
+              referenceCoef(i) - referenceMean
+          }
+        }
+        val states = optimizer.iterations(new CachedDiffFunction(costFun),
+          initialCoefficientsWithIntercept.asBreeze.toDenseVector)
+
+        /*
+           Note that in Multinomial Logistic Regression, the objective history
+           (loss + regularization) is log-likelihood which is invariant under feature
+           standardization. As a result, the objective history from optimizer is the same as the
+           one in the original space.
+         */
+        val arrayBuilder = mutable.ArrayBuilder.make[Double]
+        var state: optimizer.State = null
+        while (states.hasNext) {
+          state = states.next()
+          arrayBuilder += state.adjustedValue
+        }
+
+        if (state == null) {
+          val msg = s"${optimizer.getClass.getName} failed."
+          logError(msg)
+          throw new SparkException(msg)
+        }
+        bcFeaturesStd.destroy(blocking = false)
+
+        /*
+           The coefficients are trained in the scaled space; we're converting them back to
+           the original space.
+           Note that the intercept in scaled space and original space is the same;
+           as a result, no scaling is needed.
+         */
+        var interceptSum = 0.0
+        var coefSum = 0.0
+        val rawCoefficients = Vectors.fromBreeze(state.x)
+        val (coefMatrix, interceptVector) = rawCoefficients match {
```

Comment:
Since this code doesn't support sparse for now, maybe for readability, the following will be easier?

```scala
val rawCoefficients = state.x.toArray
val interceptVector = if ($(fitIntercept)) {
  Vectors.dense(Array.tabulate(numClasses) { i =>
    val coefIndex = (i + 1) * numFeaturesPlusIntercept - 1
    val intercept = rawCoefficients(coefIndex)
  })
} else {
  Vectors.sparse(numClasses, Seq())
}
val coefficientMatrix = Matrix.denseMatrix(numClasses, numFeatures, Array.tabulate(numClasses * numFeatures) { i =>
  // flatIndex will loop though rawCoefficients, and skip the intercept terms.
  val flatIndex = if ($(fitIntercept)) i + i / numFeatures else i
  val featureIndex = i % numFeatures
  if (featuresStd(featureIndex) != 0.0) {
    rawCoefficients(flatIndex) / featuresStd(featureIndex)
   } else {
     0.0
   }
}, isTransposed = true)
``` 

This can be another JIRA. For `Vector`, we support `compressed` api when the sparse format saves more space. `Matrix` should support it, and compress the model into sparse matrix when it's needed.